Problem: cmake build does not install malamute binaries

cmake build does not build malamute binaries, compared to the other build system with ./autogen.sh + ./configure + make + make install.

The cmake install copies the following files:

zoobab@chuchu /home/zoobab/soft/malamute/malamute/build [127]$ sudo make install
[ 80%] Built target malamute
[100%] Built target mlm_selftest
Install the project...
-- Install configuration: ""
-- Installing: /usr/local/include/mlm_library.h
-- Up-to-date: /usr/local/include/malamute.h
-- Up-to-date: /usr/local/include/mlm_proto.h
-- Up-to-date: /usr/local/include/mlm_server.h
-- Up-to-date: /usr/local/include/mlm_client.h
-- Up-to-date: /usr/local/include/mlm_msg.h
-- Up-to-date: /usr/local/include/mlm_stream_simple.h
-- Up-to-date: /usr/local/include/mlm_mailbox_simple.h
-- Up-to-date: /usr/local/include/zbits.h
-- Up-to-date: /usr/local/include/mlm_server_engine.inc
-- Up-to-date: /usr/local/include/mlm_client_engine.inc
-- Up-to-date: /usr/local/include/mlm_client_custom.xml
-- Installing: /usr/local/lib/libmalamute.a
-- Installing: /usr/local/lib/pkgconfig/libmlm.pc

While the autotools method copies more binaries:

libtool: install: /usr/bin/install -c src/.libs/malamute /usr/local/bin/malamute
libtool: install: /usr/bin/install -c src/.libs/mshell /usr/local/bin/mshell
libtool: install: /usr/bin/install -c src/.libs/mlm_tutorial /usr/local/bin/mlm_tutorial
libtool: install: /usr/bin/install -c src/.libs/mlm_perftest /usr/local/bin/mlm_perftest
 /usr/bin/install -c -m 644 include/malamute.h include/mlm_proto.h include/mlm_server.h include/mlm_client.h include/mlm_library.h '/usr/local/include'
 /usr/bin/install -c -m 644 src/libmlm.pc '/usr/local/lib/pkgconfig'
 /usr/bin/install -c -m 644 src/malamute.cfg '/usr/local/etc/malamute'
 /usr/bin/install -c -m 644 src/malamute.service '/usr/local/lib/systemd/system'

After commit #b0c6927, the cmake output is different:

==========================================================
zoobab@chuchu /home/zoobab/soft/malamute/malamute/build [16]$ cmake ..
-- The C compiler identification is GNU 4.9.3
-- The CXX compiler identification is GNU 4.9.3
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for linux/wireless.h
-- Looking for linux/wireless.h - not found
-- Looking for net/if_media.h
-- Looking for net/if_media.h - not found
-- Looking for getifaddrs
-- Looking for getifaddrs - found
-- Looking for freeifaddrs
-- Looking for freeifaddrs - found
-- Looking for include files sys/socket.h, net/if.h
-- Looking for include files sys/socket.h, net/if.h - found
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29") 
-- Checking for module 'libsodium'
--   Found libsodium, version 1.0.3
-- Found LIBSODIUM: /usr/local/lib64/libsodium.so  
-- Checking for module 'libzmq'
--   Found libzmq, version 4.2.0
-- Found ZEROMQ: /usr/local/lib64/libzmq.so  
-- Checking for module 'libczmq'
--   Found libczmq, version 3.0.2
-- Found CZMQ: /usr/lib64/libczmq.so  
CMake Error at CMakeLists.txt:178 (add_executable):
  add_executable cannot create target "malamute" because another target with
  the same name already exists.  The existing target is a shared library
  created in source directory "/home/zoobab/soft/malamute/malamute".  See
  documentation for policy CMP0002 for more details.


-- version: 0.0.0
-- install: /usr/local
-- Configuring incomplete, errors occurred!
See also "/home/zoobab/soft/malamute/malamute/build/CMakeFiles/CMakeOutput.log".
See also "/home/zoobab/soft/malamute/malamute/build/CMakeFiles/CMakeError.log".
==========================================================

If I change the line 178 to:

===========================================================
add_executable(malamute2 "${SOURCE_DIR}/src/malamute.c")
target_link_libraries(malamute2 malamute ${ZEROMQ_LIBRARIES})
set_target_properties(malamute2 PROPERTIES RUNTIME_OUTPUT_DIRECTORY "${SOURCE_DIR}/src")
===========================================================

Then it works, but the binary name is now malamute2. Any idea how to solve this properly?

I also tried to put this at the top of the CMakeLists.txt:

set(ALLOW_DUPLICATE_CUSTOM_TARGETS 1)

But it does not change anything.

--- a/.travis.yml
+++ b/.travis.yml
@@ -9,6 +9,7 @@ env:
 - BUILD_TYPE=check_zproject
 - BUILD_TYPE=check_zproto
 - BUILD_TYPE=valgrind
+- BUILD_TYPE=cmake
 
 services: 
 - docker

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -213,6 +213,9 @@ set_target_properties(
     malamute
     PROPERTIES RUNTIME_OUTPUT_DIRECTORY "${SOURCE_DIR}/src"
 )
+install(TARGETS malamute
+    RUNTIME DESTINATION bin
+)
 add_executable(
     mshell
     "${SOURCE_DIR}/src/mshell.c"
@@ -227,6 +230,9 @@ target_link_libraries(
 set_target_properties(
     mshell
     PROPERTIES RUNTIME_OUTPUT_DIRECTORY "${SOURCE_DIR}/src"
+)
+install(TARGETS mshell
+    RUNTIME DESTINATION bin
 )
 add_executable(
     mlm_tutorial
@@ -243,6 +249,9 @@ set_target_properties(
     mlm_tutorial
     PROPERTIES RUNTIME_OUTPUT_DIRECTORY "${SOURCE_DIR}/src"
 )
+install(TARGETS mlm_tutorial
+    RUNTIME DESTINATION bin
+)
 add_executable(
     mlm_perftest
     "${SOURCE_DIR}/src/mlm_perftest.c"
@@ -258,6 +267,9 @@ set_target_properties(
     mlm_perftest
     PROPERTIES RUNTIME_OUTPUT_DIRECTORY "${SOURCE_DIR}/src"
 )
+install(TARGETS mlm_perftest
+    RUNTIME DESTINATION bin
+)
 add_executable(
     mlm_selftest
     "${SOURCE_DIR}/src/mlm_selftest.c"
@@ -272,6 +284,9 @@ target_link_libraries(
 set_target_properties(
     mlm_selftest
     PROPERTIES RUNTIME_OUTPUT_DIRECTORY "${SOURCE_DIR}/src"
+)
+install(TARGETS mlm_selftest
+    RUNTIME DESTINATION bin
 )
 
 ########################################################################

--- a/bindings/jni/build.gradle
+++ b/bindings/jni/build.gradle
@@ -8,7 +8,7 @@
 plugins {
     id 'java'
     id 'maven-publish'
-    id "com.jfrog.bintray" version "1.7.2"
+    id "com.jfrog.bintray" version "1.7.3"
 }
 
 group = "org.zeromq"

--- a/builds/cmake/ci_build.sh
+++ b/builds/cmake/ci_build.sh
@@ -24,7 +24,10 @@ CMAKE_OPTS+=("-DCMAKE_INCLUDE_PATH:PATH=${BUILD_PREFIX}/include")
 if ! ((command -v dpkg-query >/dev/null 2>&1 && dpkg-query --list libzmq3-dev >/dev/null 2>&1) || \
        (command -v brew >/dev/null 2>&1 && brew ls --versions libzmq >/dev/null 2>&1)); then
     git clone --quiet --depth 1 https://github.com/zeromq/libzmq.git libzmq
+    BASE_PWD=${PWD}
     cd libzmq
+    CCACHE_BASEDIR=${PWD}
+    export CCACHE_BASEDIR
     git --no-pager log --oneline -n1
     if [ -e autogen.sh ]; then
         ./autogen.sh 2> /dev/null
@@ -43,12 +46,15 @@ if ! ((command -v dpkg-query >/dev/null 2>&1 && dpkg-query --list libzmq3-dev >/
     ./configure "${CONFIG_OPTS[@]}"
     make -j4
     make install
-    cd ..
+    cd "${BASE_PWD}"
 fi
 if ! ((command -v dpkg-query >/dev/null 2>&1 && dpkg-query --list libczmq-dev >/dev/null 2>&1) || \
        (command -v brew >/dev/null 2>&1 && brew ls --versions czmq >/dev/null 2>&1)); then
     git clone --quiet --depth 1 https://github.com/zeromq/czmq.git czmq
+    BASE_PWD=${PWD}
     cd czmq
+    CCACHE_BASEDIR=${PWD}
+    export CCACHE_BASEDIR
     git --no-pager log --oneline -n1
     if [ -e autogen.sh ]; then
         ./autogen.sh 2> /dev/null
@@ -67,11 +73,13 @@ if ! ((command -v dpkg-query >/dev/null 2>&1 && dpkg-query --list libczmq-dev >/
     ./configure "${CONFIG_OPTS[@]}"
     make -j4
     make install
-    cd ..
+    cd "${BASE_PWD}"
 fi
 
 # Build and check this project
 cd ../..
+CCACHE_BASEDIR=${PWD}
+export CCACHE_BASEDIR
 PKG_CONFIG_PATH=${BUILD_PREFIX}/lib/pkgconfig cmake "${CMAKE_OPTS[@]}" .
 make all VERBOSE=1 -j4
 ctest -V

